ako-gateway-api/k8s/gateway_controller.go

```diff
@@ -0,0 +1,585 @@
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	corev1 "k8s.io/api/core/v1"
+
+	gwApi "sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+var controllerInstance *GatewayController
+var ctrlonce sync.Once
+var tzonce sync.Once
+var transportZone string
+
+var WorkloadNamespaceCount int = 0
+var countLock sync.RWMutex
+
+type GatewayController struct {
+	worker_id uint32
+	//recorder        record.EventRecorder
+	informers        *utils.Informers
+	dynamicInformers *lib.DynamicInformers
+	workqueue        []workqueue.RateLimitingInterface
+	DisableSync      bool
+}
+
+func SharedGatewayController() *GatewayController {
+	ctrlonce.Do(func() {
+		controllerInstance = &GatewayController{
+			worker_id:        (uint32(1) << utils.NumWorkersIngestion) - 1,
+			informers:        utils.GetInformers(),
+			dynamicInformers: lib.GetDynamicInformers(),
+			DisableSync:      true,
+		}
+	})
+	return controllerInstance
+}
+
+func validateAviConfigMap(obj interface{}) (*corev1.ConfigMap, bool) {
+	configMap, ok := obj.(*corev1.ConfigMap)
+	if ok && lib.GetNamespaceToSync() != "" {
+		// AKO is running for a particular namespace, look for the Avi config map here
+		if configMap.Name == lib.AviConfigMap {
+			return configMap, true
+		}
+	} else if ok && configMap.Namespace == utils.GetAKONamespace() && configMap.Name == lib.AviConfigMap {
+		return configMap, true
+	}
+	return nil, false
+}
+
+func (c *GatewayController) Start(stopCh <-chan struct{}) {
+	go c.informers.ServiceInformer.Informer().Run(stopCh)
+	go c.informers.EpInformer.Informer().Run(stopCh)
+	go c.informers.NSInformer.Informer().Run(stopCh)
+
+	informersList := []cache.InformerSynced{
+		c.informers.EpInformer.Informer().HasSynced,
+		c.informers.ServiceInformer.Informer().HasSynced,
+		c.informers.NSInformer.Informer().HasSynced,
+	}
+
+	if !lib.AviSecretInitialized {
+		go c.informers.SecretInformer.Informer().Run(stopCh)
+		informersList = append(informersList, c.informers.SecretInformer.Informer().HasSynced)
+	}
+
+	if lib.GetServiceType() == lib.NodePortLocal {
+		go c.informers.PodInformer.Informer().Run(stopCh)
+		informersList = append(informersList, c.informers.PodInformer.Informer().HasSynced)
+	}
+
+	go gwlib.AKOControlConfig().GatewayApiInformers().GatewayClassInformer.Informer().Run(stopCh)
+	informersList = append(informersList, gwlib.AKOControlConfig().GatewayApiInformers().GatewayClassInformer.Informer().HasSynced)
+	go gwlib.AKOControlConfig().GatewayApiInformers().GatewayInformer.Informer().Run(stopCh)
+	informersList = append(informersList, gwlib.AKOControlConfig().GatewayApiInformers().GatewayInformer.Informer().HasSynced)
+
+	if !cache.WaitForCacheSync(stopCh, informersList...) {
+		runtime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+	} else {
+		utils.AviLog.Info("Caches synced")
+	}
+}
+
+func (c *GatewayController) SetupEventHandlers(k8sinfo k8s.K8sinformers) {
```

Comment:
Shall we move this to a separate file where we can keep all the eventhandler related functions (both gateway apis' and other eventhandler functions)